KAFKA-6863 Kafka clients should try to use multiple DNS resolved IP

edoardocomar opened the pull request:
* Implementation of KIP-302
* Based on the new client configuration `client.dns.lookup` a
 Selector can use InetAddress.getAllByName to find all IPs and iterates
over them when they fail to connect
* Only use IPv4 or IPv6 addresses
* Unit test

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

On attempting to connect to a resolved IP that does not respond, there appears to be a 75 second timeout - this is reflected in the unit test. 
If the resolved IP refuses the connection the selector would move the next IP quickly 

Note that we also fixed `SelectorTst.testConnectionRefused()`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

edoardocomar commented:
If we could control the timeout on the unblocking NIO connection the test could be made to pass quickly. Please feedback welcome

ijuma commented:
Thanks for the PR. This is not what I thought you were asking for. I thought you wanted the bootstrap brokers to resolve all the IPs, but the behaviour implemented is more than that. Can you elaborate on why we would want this? It seems like it could be a bit dangerous.

edoardocomar commented:
Hi @ijuma why do you think this could be dangerous ? 
The client is simply using the information returned by the DNS. 
If you don't trust your DNS to resolve your hostnames we enter a different scenario.

Why we think this JIRA/PR is useful :
Normally you'd expect a symbolic name to resolve to one IP address, but that is not necessarily the case. 
As explained in the JIRA,  our use case is for multiple load balancers fronting the Kafka cluster. The Kafka advertised listeners advertise hostnames that the DNS server maps to the IP of all LBs not just one.
If one LBs isn't available, but the client is able to use another IP for the same hostname (it connects to the 2nd LB for example) the service stays available. 

As we have not introduced any new client configturation, we thought this PR can be reviewed without a KIP. If that's not the case please let us know as we'd like this functionality to make it in v2.0.0

edoardocomar commented:
@ijuma @rajinisivaram any comments ? As said this is an internal improvment that we'd like to get in for 2.0.0
It has no impact on users whose clusters hostnames are resolved to a single IP

rajinisivaram commented:
@edoardocomar Thanks for the PR. KAFKA-6195 addresses a similar but slightly different requirement. It also started as an `internal improvement`, but it would have broken some environments. The KIP that was added later (KIP-235) was useful because in the end a config was added to enable/disable auto-lookups.

Like with KIP-235, I am not sure we want this behaviour as default. Both this PR and the KIP are targeted at multiple DNS entries and use all resolved entries. One uses (canonicalHostName, IP) and the other uses (hostName, IP). In a secure environment, host name matters. We need to clarify how these two approaches co-exist and when to use each approach.

The JIRA suggests that the use case is to resolve load balancers, which sort of points towards bootstrap list. You have `kafka.example.com` resolving to `{(kafka01.example.com, ip1), (kafka02.example.com, ip2), (kafka03.example.com, ip3)}` and want to use `kafka.example.com` in the bootstrap list. This scenario is addressed by KIP-235. Instead of using `(kafka.example.com, ip1)` as the single bootstrap server, KIP-235 changes the bootstrap list to `{(kafka01.example.com, ip1), (kafka02.example.com, ip2), (kafka03.example.com, ip3)}` .

But this PR suggests that you have more than one proxy in front of each broker for load balancing and want to resolve not just bootstrap list, but also hosts returned as advertised host names. So you have kafka01.example.com resolving to `{(kafka01.example.com, ip1a), (kafka01.example.com, ip1b)}`. Is that really the problem you are trying to solve?

In any case, I think a KIP that explains the motivation, the reason why KIP-235 is not sufficient and how the two approaches will co-exist would help. Or you could piggy-back on KIP-235 since if we have to add a config, it may be possible to adapt the config proposed in KIP-235.

edoardocomar commented:
> But this PR suggests that you have more than one proxy in front of each broker for load balancing and want to resolve not just bootstrap list, but also hosts returned as advertised host names. So you have kafka01.example.com resolving to {(kafka01.example.com, ip1a), (kafka01.example.com, ip1b)}. Is that really the problem you are trying to solve?

yes exactly

edoardocomar commented:
> I am not sure we want this behaviour as default.

well for most cases where the the DNS will resolve a name to a single IP, nothing changes

rajinisivaram commented:
@edoardocomar In some environments, something does change - hence you added the PR :-) Do all those environments expect connections to be attempted to all resolved IPs? If not, we need a config.  The default value could be what most of those environments expect as long as we don't break anyone who didn't expect this to happen (it could be to enable this change by default, just needs discussion)

As an example, I have both IPv4 and IPv6 addresses for localhost in /etc/hosts. If I run `InetAddress.getAllByName("localhost")`, I get two entries, one with an IPv4 address and another with an IPv6 address. I don't really want two connections to be attempted since one is not a fallback for the other.

Another reason why a KIP would be useful for discussion is that the PR currently uses two proxies in active/standby mode. So it is not balancing load, it is using a standby proxy for HA if connection to the active proxy fails. Do we want to randomize like we do with bootstrap brokers? Again these scenarios are easier to discuss in the context of a KIP.

edoardocomar commented:
Thanks @rajinisivaram we opened 
https://cwiki.apache.org/confluence/display/KAFKA/KIP-302+-+Enable+Kafka+clients+to+use+all+DNS+resolved+IP+addresses

edoardocomar commented:
Hi @rajinisivaram @ijuma we've overhauled the PR by implementing KIP-302 i.e. the conditional behavior based on the new config option defined by `KIP-235`

Note that we also fixed `SelectorTest.testConnectionRefused()`

mimaison commented:
@lindong28 @harshach @omkreddy @cmccabe Can one of you help with the review? Thanks

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/CommonClientConfigs.java:
```diff
@@ -93,6 +93,10 @@
                                                          + "elapses the client will resend the request if necessary or fail the request if "
                                                          + "retries are exhausted.";
 
+    public static final String CLIENT_DNS_LOOKUP_CONFIG = "client.dns.lookup";
```
@edoardocomar Since this PR is using the same config as https://github.com/apache/kafka/pull/4485, it will need to get merged on top of that right?

mimaison replied:
We copied all common files from #4485 (at the same time addressing the points we raised there) so they can be merged in any order.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/network/Selector.java:
```diff
@@ -246,17 +228,25 @@ public Selector(long connectionMaxIdleMS, int failedAuthenticationDelayMs, Metri
     @Override
-    public void connect(String id, InetSocketAddress address, int sendBufferSize, int receiveBufferSize) throws IOException {
-        ensureNotRegistered(id);
+    public void connect(Node node, int sendBufferSize, int receiveBufferSize) throws IOException {
```
Why can we iterate through addresses in NetworkClient instead of Selector?

edoardocomar replied:
Because the client calls `connect` which then hands off to NIO the asynchronous connection 
and it's in handling the keys that we can see if one IP was "bad" and all this without the client being involved at all.

It may be not the smallest change possible because of the footprint in the tests, but to us this was the change that seemed to make sense.

We are biased by the fact that the existing `Selectable.connect` signature wasn't great also for KIP-81

rajinisivaram replied:
Network layer code is already rather complex and I would avoid this type of client logic into this layer unless there was no other way. For disconnections, we already propagate channel state to clients and it would be quite easy to add another state to indicate that it may want to retry. A few advantages of doing it this way:
1. Keep network layer changes to minimum
2. We want to make this change only for clients, so the keeping the change in the client layer makes sense (making it easier to read the code when debugging broker issues)
3. Reduces the footprint of this change
4. Keeps the handling of different variations of this config in the same layer (since the other values of this config are handled in the client layer)

I haven't followed KIP-81 PR discussions, so if there are other reasons to make this change in `Selector`, then perhaps it may make sense. But just looking at this one, it feels like it should be the responsibility of `NetworkClient` to iterate through the addresses.

@hachikuji Since you reviewed the KIP-81 PR, what do you think?

edoardocomar replied:
Many thanks @rajinisivaram for taking a look under heavy load!
We see the point of not polluting the Network layer and having retries done by the client only.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/config/ClientDnsLookup.java:
```diff
@@ -0,0 +1,40 @@
+public enum ClientDnsLookup {
+
+    DEFAULT("default"),
+    USE_ALL_DNS_IPS("use_all_dns_ips");
```
The KIP says `use.all.dns.ips` and the other PR uses dot separated name.

edoardocomar replied:
yes but for consistency @mimaison commented on KIP-235 
https://github.com/apache/kafka/pull/4485#discussion_r219670881

edoardocomar replied:
of course we'd update the KIP if this change was acceptable. it breaks no one yet ...

rajinisivaram reviewed (commented):
@mimaison I think this PR is very big because it has changed the `Selectable` interface. Is there a reason why iteration through resolved addresses can't be handled in the client layer rather than the network layer?

edoardocomar committed:
KAFKA-6863 Kafka clients should try to use multiple DNS resolved IP

* Implementation of KIP-302
* Based on the new client configuration `client.dns.lookup` a
 NetworkClient can use InetAddress.getAllByName to find all IPs and
iterate over them when they fail to connect
* Only use IPv4 or IPv6 addresses
* Unit test

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

edoardocomar commented:
Hi @rajinisivaram we had another go, this time using the ClientDnsLookup configuration in `NetworkClient` and would welcome your feedback !

Please note that some of the tests added to `ClientUtilsTest` rely on the fact that `kafka.apche.org` resolves to TWO different IPs.
Convenient for us, but maybe not too reliable moving forward. We could fix that by accessing the inner class in `NodeConnectionState` directly in the unit test.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:
```diff
@@ -88,4 +92,27 @@ public static ChannelBuilder createChannelBuilder(AbstractConfig config) {
+    static List<InetAddress> resolve(String host, ClientDnsLookup clientDnsLookup) throws UnknownHostException {
+        InetAddress[] addresses = InetAddress.getAllByName(host);
+        if (ClientDnsLookup.USE_ALL_DNS_IPS == clientDnsLookup) {
+            return filterPreferedAddresses(addresses);
```
nit: `Prefered` => `Preferred`

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:
```diff
@@ -88,4 +92,27 @@ public static ChannelBuilder createChannelBuilder(AbstractConfig config) {
+    }
+
+    static List<InetAddress> filterPreferedAddresses(InetAddress[] allAddresses) {
+        List<InetAddress> preferedAddresses = new ArrayList<>();
```
as before `preferred`

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/ClientUtilsTest.java:
```diff
@@ -49,6 +54,39 @@ public void testOnlyBadHostname() {
     }
 
+    @Test
+    public void testFilterPreferedAddresses() throws UnknownHostException {
```
`Preferred`

rajinisivaram reviewed (commented):
@edoardocomar @mimaison Thanks for the updates, this looks good. Left a few minor comments (just nits). Also need to rebase to resolve conflicts. I think the only issue remaining to merge this is the naming of the config value (which is currently different in PR #4485). If that is agreed today, then we can merge this to 2.1.0 and trunk.

edoardocomar committed:
KAFKA-6863 Kafka clients should try to use multiple DNS resolved IP

fixed spelling as per review

Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>
Co-authored-by: Mickael Maison <mickael.maison@gmail.com>

mimaison commented:
@rajinisivaram Thanks for the review. We've rebased on trunk and fixed the typo!

rajinisivaram reviewed (approved):
@edoardocomar @mimaison Thanks for the updates, LGTM. As @lindong28 has agreed for this to go into 2.1.0, I will merge to 2.1.0 and trunk after the builds complete.

edoardocomar commented:
Thanks - will update the KIP

rajinisivaram commented:
Test failures look unrelated, merging to trunk and 2.1.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/config/ClientDnsLookup.java:
```diff
@@ -0,0 +1,40 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.config;
```
@edoardocomar @mimaison `org.apache.kafka.common.config` is a public package and this is not a public API as-per the KIPs. I dont think we include similar enums in the config package. Since this is used only for client-side with `NetworkClient`, we should move `ClientDnsLookup` into `org.apache.kafka.clients` as it is in the other PR. Can you submit a follow-on PR to fix this?

mimaison replied:
Sure, we'll do that now

mimaison replied:
I see they just made that change in #4485.

rajinisivaram replied:
@mimaison thanks, it can be updated in that PR if they find it easier.
